benches/benches/bevy_ecs/iteration/batched_compute.rs

```diff
@@ -0,0 +1,238 @@
+use bevy_ecs::prelude::*;
+use core::arch::x86_64::*;
+use glam::*;
+use rand::prelude::*;
+
+use criterion::BenchmarkId;
+use criterion::Criterion;
+
+#[derive(Component, Copy, Clone, Default)]
+struct Position(Vec3);
+
+#[derive(Component, Copy, Clone, Default)]
+#[repr(transparent)]
+struct Health(f32);
+
+//A hyperplane describing solid geometry, (x,y,z) = n with d such that nx + d = 0
+#[derive(Component, Copy, Clone, Default)]
+struct Wall(Vec3, f32);
+
+struct Benchmark(World);
+
+fn rnd_vec3(rng: &mut ThreadRng) -> Vec3 {
+    let x1 = rng.gen_range(-16.0..=16.0);
+    let x2 = rng.gen_range(-16.0..=16.0);
+    let x3 = rng.gen_range(-16.0..=16.0);
+
+    Vec3::new(x1, x2, x3)
+}
+
+fn rnd_wall(rng: &mut ThreadRng) -> Wall {
+    let d = rng.gen_range(-16.0..=16.0);
+
+    Wall(rnd_vec3(rng).normalize_or_zero(), d)
+}
+
+// AoS to SoA data layout conversion for x86 AVX.
+// This code has been adapted from:
+//   https://www.intel.com/content/dam/develop/external/us/en/documents/normvec-181650.pdf
+#[inline(always)]
+// This example is written in a way that benefits from inlined data layout conversion.
+fn aos_to_soa_83(aos_inner: &[Vec3; 8]) -> [__m256; 3] {
+    unsafe {
+        //# SAFETY: Vec3 is repr(C) for x86_64
+        let mx0 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(0));
+        let mx1 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(4));
+        let mx2 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(8));
+        let mx3 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(12));
+        let mx4 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(16));
+        let mx5 = _mm_loadu_ps((aos_inner as *const Vec3 as *const f32).offset(20));
```

Comment:
If this is `glam::Vec3`, this is unsound and `repr(C)` is not enough. Vec3, as opposed to Vec3A, has padding bytes. That is, its layout is:

| f32 | f32 | f32 |  ? |
| --- | --- | --- | --- |
| init | init | init | *poison* |

so this is filling the 4th lane with data that should not be read. Vec3A is defined as `__m128` already. You must use a full 16-byte type with *zero* padding, consistently.

Reply:
Thanks for the review!  Indeed, you're right: the C (and C++) standards don't provide any guarantee against arbitrary padding being added between elements or at the end of a struct, even in the case it wouldn't be necessary to guarantee stride alignment in sequential array accesses.  It's perfectly legitimate for a conforming C compiler to lay `glam::Vec3` out in exactly the way you have suggested, including using a trap representation for the padding at the end of the struct.  Even if `glam::Vec3` were defined in terms of a single array `[f32; 3]`, you still can't escape padding at the end as you have indicated, unless you made it `repr(transparent)` anyway.

In C++, the way we'd normally deal with this is with a static assert for `sizeof(Vec3) == 3*sizeof(float)`.  It's a hack, but it works to avoid this problem.  Indeed, on all major x86 ABIs, this holds.  Since static asserts don't really exist in Rust, I'm not sure how to accomplish something equivalent there in this example.

I'd like to mention though, that real x86 C (and C++) compilers would never, ever add padding for the sake of adding it.  Padding is something done on x86 to ensure alignment, and the alignment of `Vec3` is not required to be anything higher than the alignment of `f32`.  It *could* be, and that would be legal in a theoretic sense for C, but no implementation out there -- on x86 -- would ever do it.  For one thing, that would contradict the psABI and break compatibility with other code on x86 systems.  However yes, one could construct a toy C compiler that would do just that, say to anticipate vectorization later on -- although I expect such a transform would be a pessimization a lot more often than not.

So, laying out `Vec3` with 12 bytes of space and aligning to the same alignment of `f32` meets the alignment requirements just fine -- and this is what real implementations -- on x86 -- do.  Not that this excuses anything, mind you, because I agree: the code should be correct from a theoretic sense, or guarded by a compile-time guard to ward off the UB parts (as one would do in C++).

With all this said, as someone that isn't super in the know of the current Rust landscape in dealing with these things, what's the standard practice here?  In C++ I'd stick a static assert on it and be done with it.  How do we make this work in Rust?

Reply:
Hmm, it was very late at night when I scanned through things here, and I both mislaid some mental math while reading this and flubbed what I meant to say. My apologies. Let me correct myself: the safety documentation here needs to explicate the actual safety precondition which is the layout match of `8 * 3 / (6 * 4) == 1` and these are unaligned loads, instead of a handwave about `repr(C)`, as if `(n * x) / (m * y) == 1` doesn't hold, **then** you have the possibility of reading the poisoned value.

You needn't particularly worry about toy C compilers, and I wasn't trying to language lawyer you on those. Rust picks specific C compilers to be compatible with (though it gets more vague if e.g. two compilers emit "the same" ABI but actually disagree on fine details). Thus when targeting x86_64-pc-windows-msvc, obviously it emits compatible code with the Windows x64 ABI as emitted by MSVC, etc. etc. down the line.

You can generate a "static assert" in Rust using the pattern
```rust
const _: () = {
    let precondition = {
        // some code that can run in const eval
    };
    if precondition == false {
        panic!("a nicely-written message")
    };
};
```

Reply:
No problem and thanks for the feedback.  I wrote a good amount of this PR myself during some late nights 😅  It sounds like we're converging on a solution.  Would it be acceptable to document this about the expected layout of Vec3 for x86 in the example, and use that pattern to statically assert this is the case for completeness?  With maybe an additional comment about how a `repr(transparent)` type would be preferable for code that needs to be portable?

Reply:
Discussing `repr(transparent)` isn't needed but otherwise yes, that sounds fine. My real concern is just about someone retouching this code later without a deep and horrifying knowledge we have of the details of the x86-64 ABI, what `__m128` even is, and C layout, in a way that changes the numbers and causes the the last `_mm_loadu_ps` to read 1 more float than actually exists.